### ISSUE TITLE
Material n()/k() caches become stale after parameter updates

### ISSUE THREAD

**Issue by mabl** [optiland/optiland#794]
# Material n()/k() caches become stale after parameter updates

## Summary
After `n()` or `k()` has populated a material's wavelength cache, replacing a
parameter can leave cached values from the previous material state. In the
formula-based example below, a new wavelength uses the updated coefficients,
while a previously queried wavelength returns the old value. With Torch, warming
the cache before replacing an `IdealMaterial` parameter with a new
`nn.Parameter` can also return a detached cached result and prevent gradient flow.

This reproduces for `IdealMaterial.index` / `absorp` and for formula-based
`Material.coefficients`, which is provided through `MaterialFile`.
Before proposing a fix, I would like guidance on the intended
parameter-update contract because invalidation is a shared concern rather than
an `IdealMaterial`-specific issue.

**A cached property belongs to a particular material state, and an update must
not reuse results or gradient connections from the previous state.**

## Checklist
- [x] Searched related issues and discussions, including #484, #485, and #630.
- [x] Read the material APIs, cache implementation, and existing parameter-gradient tests.
- [x] Reproduced on upstream `master` at commit `7df37590c2d561d95403dc264e27b8b9a77ee879`.
- [x] Included fresh-instance and trainable-before-first-evaluation controls.

## Numerical Reproducer
This uses public attributes and the bundled N-BK7 catalog data. The coefficient
change is synthetic: it isolates cache behavior, not a proposed physical glass
model. Wavelengths are in micrometers.

```python
import numpy as np

import optiland.backend as be
from optiland.materials import IdealMaterial, Material


be.set_backend("numpy")
be.set_precision("float64")

ideal = IdealMaterial(1.33, 0.02)
print("ideal before:", ideal.n(0.55), ideal.k(0.55))
ideal.index = be.array([1.5])
ideal.absorp = be.array([0.04])
fresh_ideal = IdealMaterial(1.5, 0.04)
print("ideal after: ", ideal.n(0.55), ideal.k(0.55))
print("ideal fresh: ", fresh_ideal.n(0.55), fresh_ideal.k(0.55))

glass = Material("N-BK7", catalog="schott")
fresh_glass = Material("N-BK7", catalog="schott")
print("glass before:", glass.n(0.55).item())
coefficients = glass.coefficients.copy()
coefficients[0] += 0.1
glass.coefficients = coefficients
fresh_glass.coefficients = coefficients.copy()
print("glass after: ", glass.n(0.55).item())
print("glass fresh: ", fresh_glass.n(0.55).item())

# An unqueried wavelength sees the new coefficients on both objects.
assert np.allclose(glass.n(0.60), fresh_glass.n(0.60), rtol=1e-12, atol=0.0)
```

Upstream results at the previously queried wavelength `0.55`:

| Property | After replacement | Fresh control |
| --- | ---: | ---: |
| Ideal `n` | 1.33 | 1.5 |
| Ideal `k` | 0.02 | 0.04 |
| N-BK7 `n` | 1.5185223876207927 | 1.551099687868434 |

Expected: each edited object agrees with its fresh control. For N-BK7, both
objects return `1.5489189777859838` at the previously unqueried wavelength `0.60`.
The edited coefficients are being read, but not on the old cache-hit path.

## Gradient Consequence
Numerical equality does not make a cached result valid for a new trainable leaf.
This standalone example changes only whether the material was evaluated before
installing the parameter:

```python
import torch

import optiland.backend as be
from optiland.materials import IdealMaterial


be.set_backend("torch")
be.set_device("cpu")
be.set_precision("float64")
for prop, attribute in (("n", "index"), ("k", "absorp")):
    for warm in (True, False):
        be.grad_mode.disable()
        with torch.no_grad():
            material = IdealMaterial(1.33, 0.02)
            wavelength = be.array([0.45, 0.55, 0.65])
            if warm:
                getattr(material, prop)(wavelength)
            leaf = torch.nn.Parameter(getattr(material, attribute).clone())
            setattr(material, attribute, leaf)

        be.grad_mode.enable()
        with torch.enable_grad():
            result = getattr(material, prop)(wavelength)
            print(prop, "warm:", warm, "requires_grad:", result.requires_grad)
            try:
                result.square().sum().backward()
            except RuntimeError as error:
                print("backward:", error)
            print("parameter gradient:", leaf.grad)
```

Upstream, both warmed cases return `requires_grad=False`; backward raises
`element 0 of tensors does not require grad and does not have a grad_fn`, and
the parameter gradient remains `None`. Both unwarmed controls succeed, with
gradients `7.98` for `index` and `0.12` for `absorp`. Expected: both histories
produce a result connected to the current leaf and those same gradients.

The new property call occurs **with gradients enabled**. This is not an attempt
to differentiate a result retained from inside `no_grad()`.

## Existing Contract And Cause
These attributes are not all private implementation details:

- [`MaterialFile.coefficients`](https://github.com/optiland/optiland/blob/7df37590c2d561d95403dc264e27b8b9a77ee879/optiland/materials/material_file.py#L22-L41) is a documented public attribute, inherited by `Material`.
- Existing [gradient tests](https://github.com/optiland/optiland/blob/7df37590c2d561d95403dc264e27b8b9a77ee879/tests/test_materials.py#L436-L478) directly assign trainable tensors to `IdealMaterial.index` and `absorp`.
- The built-in [`set_index` updater](https://github.com/optiland/optiland/blob/7df37590c2d561d95403dc264e27b8b9a77ee879/optiland/optic/optic_updater.py#L113-L124) creates a new material. Its normal index-optimization path is not the same-instance failure demonstrated here.

[`BaseMaterial`](https://github.com/optiland/optiland/blob/7df37590c2d561d95403dc264e27b8b9a77ee879/optiland/materials/base.py#L213-L365)
keys these caches by wavelength and kwargs, not material parameters. The
assignments above do not invalidate the caches. A hit returns before current
parameters are evaluated or the result-gradient check runs.

## API Direction For Discussion
**What parameter-mutation contract should material caches support: explicit
updates with invalidation, automatic invalidation for supported assignment or
mutation paths, or another state-versioning approach?**

Immutable numerical parameter snapshots with explicit update methods are one
possible direction. Updates could validate new values and invalidate dependent
state, while unchanged materials keep their caches without inspecting all
parameter contents on each lookup. A read-only Python property alone would not
make array or tensor storage immutable.

Before implementing this, I would appreciate guidance on:

1. **Compatibility:** Is deprecating direct parameter assignment and in-place edits acceptable, or should those operations continue to work with automatic invalidation?
2. **Update semantics:** Should an update return a new material, or replace immutable parameter values inside the existing material and invalidate its caches? This matters when multiple surfaces share a material instance.
3. **Trainable tensors:** How should the contract accommodate optimizer updates to `torch.nn.Parameter` in place? Would immutable numerical snapshots and a separate trainable-state policy be preferable? Caller-owned tensor aliases and property-result views can still expose writable storage, and protecting state must not silently detach gradients or replace the optimizer's leaf.

A shared update contract would address more than ideal constants. Simply
disabling `IdealMaterial` caching would leave the dispersion-coefficient case
above unresolved. I am happy to contribute a fix and regression tests once the
intended API and compatibility requirements are agreed; no particular method
names or foundational implementation are proposed here.

## Scope And Related Work
The demonstrated nonideal failure is for `n()`, not a claim that changing
dispersion coefficients affects `k()`. Ordinary wavelength and scalar
temperature/pressure queries on unchanged materials are not shown to be broken.
Some in-place edits also reproduce stale results, but replacement alone
establishes the defect above.

Closed [#484](https://github.com/optiland/optiland/issues/484), fixed by
[#485](https://github.com/optiland/optiland/pull/485), concerned reusing an
already-consumed backward graph. Here the cached numerical result was never
connected to the new leaf.
[#630](https://github.com/optiland/optiland/issues/630) concerned wavelength-array
cache-key collisions; these examples do not require different wavelength arrays
or allocator-address reuse.

## Environment
Windows, Python 3.14.2, NumPy 2.5.3, Torch 2.13.0 CPU; float64 in the examples.
No accelerator behavior is claimed.
